src/dotty/tools/dotc/core/pickling/TreeBuffer.scala

```diff
@@ -0,0 +1,163 @@
+package dotty.tools
+package dotc
+package core
+package pickling
+
+import util.Util.{bestFit, dble}
+import TastyBuffer.{Addr, AddrWidth}
+import config.Printers.pickling
+
+class TreeBuffer extends TastyBuffer(1000000) {
```

Comment:
TreeBuffer should have a comment at the top explaining the overall strategy of how offsets are first encoded simply and wastefully, and later squeezed to save space.